### ISSUE TITLE
Investigating memory corruption at Illien

### ISSUE THREAD

**Issue by Hexaae** [Pyrdacor/Ambermoon#66]
Exit and entering Pelanis palace a couple of time after I fly with Giant Eagle to Illien can randomly cause strange gfx corruption like this:
![WinUAE Screenshot 2022 08 17 - 18 36 26 53](https://user-images.githubusercontent.com/8591462/185197136-7f28b19d-d863-4462-b274-41d1b0636818.png)
Sometimes the walls have the wrong tile with windows or plaques at every wall tile (???).

Still investigating a reproducible pattern...

**Comment by Hexaae** [comment]
EDIT: it seems the bug is reproducible entering/exiting Pelanis Palace/Illien multiple times, but only when using the WHDLoad slave 2.1

**Comment by Pyrdacor** [comment]
Then you might tell Wepl about it.

**Comment by Hexaae** [comment]
Reported in your WHD thread about Ambermoon 1.16 release...

**Comment by Hexaae** [comment]
Oh no, I had a strange issue also running 1.16 from WB: when sleeping I got msgs truncated like "Luca0150" instead of usual "Luca regained 150 magic points...", it could still be something in 1.1.6

**Comment by Pyrdacor** [comment]
Can you please open a new issue for the text stuff?

**Comment by Hexaae** [comment]
> Can you please open a new issue for the text stuff?

Let me investigate if I can reproduce it first, I'll also degrade emu speed etc.

**Comment by Hexaae** [comment]
ATM all I can say is that it's reproducible even with the original game 1.08

**Comment by a1exh** [comment]
V1.07 and below were the originals

**Comment by Hexaae** [comment]
Ok. Tested up to 1.13: my savegame above entering Illien twice solves the problem (sort of fixes itself). With 1.14+ I can reproduce it every time.

**Comment by Hexaae** [comment]
To me it seems an original game issue when entering Illien (376,277) after a trip with the Eagle to/from Castle Godsbane...
Can you please try to reproduce the strange textures corruptions in Illien? I could be a savegame issue, but it sounds like an original game bug at this point...

**Comment by Hexaae** [comment]
Started a new game with 1.16 in the meanwhile and, using cheat with Netsrak's Staff, I got the flute to call the Giant Eagle to fly to the nearest town: I found NewLake and I noticed the same issue with (wrong) wall textures the first time I entered the gate!
It seems a generic issue with 3D view in towns...

**Comment by Hexaae** [comment]
It is an original game bug anyway, reproducible not only in Illien but also in other towns: after a long trip with the giant eagle it often occurs that the 3D city textures are incorrect. Not only an 060 bug: on WinUAE it happens also with 030 or 040 emu.

**Comment by a1exh** [comment]
Not that it matters but the original games were 1.05 (DE) and 1.07 (Eng).

Bugs like this have been reported in the past but no-one ever posted real technical information that could be debugged.

e.g. Someone called st-th (who is now unreachable) posted some tiny tiny fractions of details.

http://eab.abime.net/showpost.php?p=531661&postcount=174

**Comment by Hexaae** [comment]
Ouch! So there are a lot of code bugs indeed... I love this game 😫

**Comment by a1exh** [comment]
Maybe. He never published his work. But some of his posts were extremely technical and led to bug fixes. 
Such as the herbs that never worked

http://eab.abime.net/showpost.php?p=891734&postcount=516

**Comment by Pyrdacor** [comment]
> Maybe. He never published his work. But some of his posts were extremely technical and led to bug fixes. Such as the herbs that never worked
> 
> http://eab.abime.net/showpost.php?p=891734&postcount=516

I wish I had that information back then. I had to find those two bugs on my own. His item slot format description is not 100% correct but it's amazing to see how much knowledge about the game formats he had back then. It's a shame he never published it.

Today I guess I am aware of most bugs outside of the executables and know all the file formats in detail.

While I develop AA, I also found some more bugs in code and also decoded more bits and bytes of the file formats. But most of the bugs are minor or only lead to problems with additional content.

Just recently I found one. The is the spell Blink which is the only spell in the game which targets a free battle field. All other spells target monsters, players, items or nothing at all.

Such spells as Blink are stored in a list of pointers with some additional data. Something like "FieldTargetedSpells". Each entry starts with a long, where the first word is the spell school and the second word is the spell index. Then follows a pointer to the function which does the target selection routine. Then some code loops over the list and checks for a spell match. There is an end marker of 0xffff in the spell list. However the code which loops checks for 0xffffffff. So in case it couldn't find the spell for some reason, it would interpret the word 0xffff as the spell school and the next word (following data) as the spell index. This won't match either so it keeps going until it finds a 0xffffffff by luck. This shouldn't hurt, but still is a bug. The code is only used for the spells with that target type so only for Blink. And of course it will match instantly. If someone sets the target type of a spell to "field" and don't add the table entries correctly, the code may search through the whole memory in worst case and eventually even crash when reaching the memory boundary etc.

**Comment by a1exh** [comment]
> I wish I had that information back then. I had to find those two bugs on my own.

Perhaps you discovered it before we started working together. This bug was in my spreadsheet in full detail including a link to that page . You even cross reference my spreadsheet entry (Unfixed 46) when you committed the change.

https://github.com/Pyrdacor/Ambermoon/commit/16f305b82f567df5b117d2c2aa123d4430ed41e8#diff-970d9d07a7064d52efff7f30b4c4f0bfdd786a721e4c1a1d7e83fa97e12805af

**Comment by Pyrdacor** [comment]
Yeah I knew the bug from your spreadsheet of course but most likely overlooked the link. ;)

Most of the times, the link only contained the description of the bug. I had no clue that someone really posted some details about the actual fix or even description of code errors. So maybe I just skipped that link which turned out to be a mistake now.

**Comment by Hexaae** [comment]
Just as a reminder and backlink to keep track of these threads:
http://mantis.whdload.de/view.php?id=5716
https://imgsli.com/MTIxMzcy example of random corrupt textures when exiting outdoor...

**Comment by Pyrdacor** [comment]
Is the texture issue temporary? How long does it occur?

**Comment by Hexaae** [comment]
The pattern to reproduce ithis issue is:
- call the eagle and fly to castle Godsbane. Enter the castle and open the map.
- exit the castle and fly to Illien or another big city. Enter the city and enter f.e. Pelanis palace. Then exit the palace.
At this point you can usually notice strange texture corruption in the city outdoor...

**Comment by Pyrdacor** [comment]
I mean: is the texture changing while you move on the map or is it constant as long as you be on the same map?

**Comment by Pyrdacor** [comment]
It looks more like some tiles have the wrong texture index so the map data is somewhat broken.

**Comment by Hexaae** [comment]
It happens randomly anyway and only following the instruction. If I enter/exit Illien city 10 times in a row it won't happen. If I move to another town flying there with the eagle, and then go back to Illien there is a higher chance to see the bug...

**Comment by Pyrdacor** [comment]
I hope my memory doesn't fool me here but in Ambermoon Advanced I think there were some issues as well. Also when changing some maps a few times. It was also a 3D map and you enter it from a "world map". You also enter Illien from the world map so this might be related.

World maps have a special flag bit set. I think the one flag is important, which controls how the 2D player is drawn. This changes a lot of stuff in the original code so there might be a specific bug when you move from a "world map" to a 3D map.

In the original code there is a "map type" variable. While the data only specify 2D and 3D as a type, the game differentiates 4 types:

- World map (2D)
- Other 2D map
- City (3D with sky)
- 3D dungeon

The main reason for splitting them is drawing I guess but also some changes in logic.

Here a change from the first one to the third one happen. And I think exactly that transition might have a flaw. But it will be hard to find it.

**Comment by Pyrdacor** [comment]
It would be interesting if the bug is persistent when it once happens. If it occurs, can you try to revert or change it by moving again to another town and back?

Moreover does it affect the whole town of Illien or only some parts?

From the screenshots it is in the lower part of the map. Would be interesting if it also happens in the top part as well.

**Comment by Hexaae** [comment]
It's random (but only after some patterns described above, or at least that is the easy way...) and not persistent: just entering shops/houses in the town and exiting usually fixes it...
Whole town uses the same few textures so yes, when it happens is extended to the whole town...

Will check for the N vs S map towns but I don't think will change.,,

**Comment by Pyrdacor** [comment]
This sounds like wrong branching or memory access during map load then. As only some tiles are affected, I think it can't be no major memory error but is most likely a slightly incorrect index calculation.

Would be interesting to have a 1:1 comparison of changed tiles. For example: is very wall of type X after the bug of type Y? Or is it totally random? This would provide some clues in which data loading the error happens.

The data is 2 part. The lab data which stored the textures for walls and objects (and other stuff) and the map itself which references those walls and objects via index.

So either the index on the map is wrong or the texture index in the lab data (or just the wrong gfx is loaded into memory for that index).

**Comment by Hexaae** [comment]
It's always the same corrupt appearance for all instances (X -> Y), and identical everywhere is used.
It usually happens for the walls. Rarely it happens for the floor as in the first screenshot, and usually it doesn't look "corrupt fastmem loaded into chip"-like but just a replacement with the wrong one (which looks "good", no corruption except you notice is not the correct texture...).

Following my [repro-steps above](https://github.com/Pyrdacor/Ambermoon/issues/66#issuecomment-1221542019), with my savegame everyone should be able to reproduce it.

**Comment by Pyrdacor** [comment]
I will have a closer look at it soon. Thanks for investigating.

**Comment by Hexaae** [comment]
News?

**Comment by Hexaae** [comment]
This should be easy to reproduce: enter Illien, and notice columns. Eneter Palanis Palace and exit. Look at how columns change to a different texture:

<a href="https://ibb.co/LYPwLSN"><img src="https://i.ibb.co/BcPDXLn/Illien-bug.gif" alt="Illien-bug" border="0"></a>

**Comment by Pyrdacor** [comment]
So I am not sure what those strange noise on the floor is in your first screenshot. But regarding the wall changes:

Walls can have overlays. Overlays are used for windows, pillars, signs and so on. As the overall map appearance is the same but only some details are missing, I really expect that the overlays are not loaded properly.

The overlays are missing when you enter Illien (a 3D map) from the world map. And they are there if you enter from a normal 2D map (Pelanis' palace).

So I think changing from world map to 3D city map might cause issues with overlay loading.

**Comment by Pyrdacor** [comment]
Btw this seems to be a general problem. If I enter Newlake from the world map I see riddle mouths on all walls (this is also an overlay but the wrong one). Entering the tavern and back to the city will correct the overlays.

When I leave and re-enter it is alternates between "wrong overlays/no overlays" and "correct overlays".

**Comment by Pyrdacor** [comment]
What I can tell from code:

There is a general purpose buffer (piece of memory) which is used for all kinds of things.

But interestingly it is used to store the overlay wall information temporarily and it also stores map indices of adjacent world map pieces. So maybe when leaving a world map it clears that data while the new 3D map already filled it with the overlay information. I have to dive deeper.
